src/actions/languages/python/motion.ts

```diff
@@ -0,0 +1,256 @@
+ * using their relative locations to make decisions.
+ */
+export class PythonDocument {
+  _document: TextDocument;
+  structure: StructureElement[];
+
+  constructor(document: TextDocument) {
+    this._document = document;
+    const parsed = PythonDocument._parseLines(document);
+    this.structure = PythonDocument._parseStructure(parsed);
+  }
+
+  /*
+   * Generator of the lines of text in the document
+   */
+  static *lines(document: TextDocument): Generator<string> {
+    for (let index = 0; index < document.lineCount; index++) {
+      yield document.lineAt(index).text;
+    }
+  }
+
+  /*
+   * Calculate the indentation of a line of text.
+   * Lines consisting entirely of whitespace of "starting" with a comment are defined
+   * to have an indentation of "null".
+   */
+  static _indentation(line: string): number | null {
+    const index: number = line.search(/\S/);
+
+    // Return null if line is empty, just whitespace, or starts with a comment
+    if (index === -1 || line[index] === '#') {
+      return null;
+    }
+
+    return index;
+  }
+
+  /*
+   * Parse a line of text to extract LineInfo
+   * Return null if the line is empty or starts with a comment
+   */
+  static _parseLine(index: number, text: string): LineInfo | null {
+    const indentation = this._indentation(text);
+
+    // Since indentation === 0 is a valid result we need to check for null explicitly
+    return indentation !== null ? { line: index, indentation, text } : null;
+  }
+
+  static _parseLines(document: TextDocument): LineInfo[] {
+    const lines = [...this.lines(document)]; // convert generator to Array
+    const infos = lines.map((text, index) => this._parseLine(index, text));
+
+    return infos.filter((x) => x) as LineInfo[]; // filter out empty/comment lines (null info)
+  }
+
+  static _parseStructure(lines: LineInfo[]): StructureElement[] {
+    const last = lines.length;
+    const structure: StructureElement[] = [];
+
+    for (let index = 0; index < last; index++) {
+      const info = lines[index];
+      const text = info.text;
+
+      if (text.match(/^\s*(def|class) /)) {
+        const type = text.match(/def/) ? 'function' : 'class';
+
+        // Find the end of the current function/class
+        let idx = index + 1;
+
+        for (; idx < last; idx++) {
+          if (lines[idx].indentation <= info.indentation) {
+            break;
+          }
+        }
+
+        // Since we stop when we find the first line with a less indentation
+        // we pull back one line to get to the end of the function/class
+        idx--;
+
+        const endLine = lines[idx];
+
+        structure.push({
+          type,
+          start: {
+            line: info.line,
+            character: info.indentation,
+          },
+          end: {
+            line: endLine.line,
+            character: endLine.text.search(/(?<=\S)\s*$/) - 1, // Calculate position of last non-white character
```

Comment:
I believe inline regexes are parsed each time they're used - consider making this a private static property to avoid incurring that cost for each line

Reply:
Good call. Done.